lib/IRGen/GenDecl.cpp

```diff
@@ -2868,8 +2854,8 @@ llvm::GlobalValue *IRGenModule::defineAlias(LinkEntity entity,
   auto *alias = llvm::GlobalAlias::create(
       ptrTy->getElementType(), ptrTy->getAddressSpace(), link.getLinkage(),
       link.getName(), definition, &Module);
-  alias->setVisibility(link.getVisibility());
-  alias->setDLLStorageClass(link.getDLLStorage());
+  ApplyIRLinkage({link.getLinkage(), link.getVisibility(), link.getDLLStorage()})
+      .to(alias);
```

Comment:
This pattern — repeated over and over again — where we create a declaration and then immediately override half the values from it with the values of an `IRLinkage` is quite unfortunate.  I like the idea of encapsulating `IRLinkage`; perhaps it ought to have functions on it to actually create these globals?

Reply:
I disliked that we were setting the properties for the linkage everywhere but didn't manage to do it very uniformly.  So, the application of the IRLinkage at least ensures we set the properties more uniformly.  But, I can try to sink this into createVariable and createFunction.  I'm happy to try to do that in a follow up.

Reply:
I'm fine with doing it in multiple stages.  Some other suggestions:

- As mentioned, please add functions to create variables and functions.
- Please look around IRGen for other places that create variables and functions that could use this.  LLVM's API for creating globals is really a mess, so it shouldn't be hard to create something that's friendlier and therefore encourages people to use it (and thus implicitly get the correctness benefits).
- Please add static factories (like `shared()` or `internal()`) for the common patterns.
- Consider incorporating `unnamed_addr` into this, and maybe an optional section name.  But `constant` probably ought to be a mandatory argument to the function to create a global, not part of the linkage.

Reply:
Yeah, `unnamed_addr` and `dso_local` are other attributes that I think that we should grow into this.  But, this unblocks Windows builds and improves things.